File: devops/github-actions-deep-dive/final-project/scripts/check.js
const fs = require('fs')
const path = require('path')

const root = path.resolve(__dirname, '..', 'cicd-pipeline')

const REQUIRED_FILES = [
  '.github/workflows/lint.yml',
  '.github/workflows/test.yml',
  '.github/workflows/build.yml',
  '.github/workflows/security.yml',
  '.github/workflows/deploy.yml',
  'Dockerfile',
  'package.json',
  'README.md',
]

const REQUIRED_DIRS = [
  '.github',
  '.github/workflows',
  'src',
  'tests',
]

let passed = 0
let failed = 0

function check(desc, condition) {
  if (condition) {
    console.log(`  ✅ ${desc}`)
    passed++
  } else {
    console.log(`  ❌ ${desc}`)
    failed++
  }
}

function fileExists(rel) {
  return fs.existsSync(path.join(root, rel))
}

function dirExists(rel) {
  const p = path.join(root, rel)
  return fs.existsSync(p) && fs.statSync(p).isDirectory()
}

function fileContains(rel, text) {
  if (!fileExists(rel)) return false
  return fs.readFileSync(path.join(root, rel), 'utf-8').includes(text)
}

console.log('🔍 CI/CD 流水线 — 结构验证\n')

if (!fs.existsSync(root)) {
  console.log('❌ 项目目录不存在:', root)
  console.log('\n请先创建项目：')
  console.log('  mkdir cicd-pipeline && cd cicd-pipeline && npm init -y')
  process.exit(1)
}

console.log('📁 目录结构：')
REQUIRED_DIRS.forEach((d) => check(`目录 ${d} 存在`, dirExists(d)))

console.log('\n📄 必需文件：')
REQUIRED_FILES.forEach((f) => check(`文件 ${f} 存在`, fileExists(f)))

console.log('\n🔧 工作流内容检查：')
check('lint.yml 包含 on.push 或 on.pull_request', fileContains('.github/workflows/lint.yml', 'on:'))
check('test.yml 包含 matrix 策略', fileContains('.github/workflows/test.yml', 'matrix'))
check('build.yml 包含 docker 相关步骤', fileContains('.github/workflows/build.yml', 'docker'))
check('security.yml 包含扫描步骤', fileContains('.github/workflows/security.yml', 'scan') || fileContains('.github/workflows/security.yml', 'trivy') || fileContains('.github/workflows/security.yml', 'codeql'))
check('deploy.yml 包含 environment 配置', fileContains('.github/workflows/deploy.yml', 'environment'))

console.log('\n📦 项目文件：')
check('Dockerfile 存在', fileExists('Dockerfile'))

console.log('\n📦 依赖检查：')
const pkgPath = path.join(root, 'package.json')
if (fileExists('package.json')) {
  const pkg = JSON.parse(fs.readFileSync(pkgPath, 'utf-8'))
  check('package.json 有 name 字段', !!pkg.name)
  check('package.json 有 scripts 字段', !!pkg.scripts)
}

console.log(`\n📊 结果：${passed} 通过，${failed} 失败`)
if (failed > 0) process.exit(1)

File: ai-specialization/ai-code-migration/final-project/scripts/check.py
import os
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
passed = 0
failed = 0


def check(description: str, condition: bool):
    global passed, failed
    if condition:
        print(f"  ✅ {description}")
        passed += 1
    else:
        print(f"  ❌ {description}")
        failed += 1


def file_exists(rel: str) -> bool:
    return (ROOT / rel).is_file()


def dir_exists(rel: str) -> bool:
    return (ROOT / rel).is_dir()


def file_contains(rel: str, pattern: str) -> bool:
    p = ROOT / rel
    if not p.is_file():
        return False
    content = p.read_text(encoding="utf-8")
    return pattern in content


print("\n🔍 遗留系统迁移工具链 - 项目验证\n")

# ── 1. 项目基础文件 ──
print("📁 项目基础文件：")
check("README.md 存在", file_exists("README.md"))
check("docker-compose.yml 存在", file_exists("docker-compose.yml"))
check(".env.example 存在", file_exists(".env.example"))

# ── 2. 前端结构 ──
print("\n📁 前端结构：")
check("frontend/ 目录存在", dir_exists("frontend"))
check("frontend/package.json 存在", file_exists("frontend/package.json"))
check("frontend/src/ 目录存在", dir_exists("frontend/src"))

# ── 3. 后端结构 ──
print("\n📁 后端结构：")
check("backend/ 目录存在", dir_exists("backend"))
check("backend/src/ 目录存在", dir_exists("backend/src"))
check("backend/package.json 存在", file_exists("backend/package.json"))

# ── 4. 核心模块 ──
print("\n🔧 核心模块：")
check("代码分析器存在", dir_exists("backend/src/analyzer"))
check("迁移规划器存在", dir_exists("backend/src/planner"))
check("自动转换器存在", dir_exists("backend/src/converter"))
check("测试生成器存在", dir_exists("backend/src/tester"))
check("REST API 存在", dir_exists("backend/src/api"))

# ── 5. 转换规则 ──
print("\n📐 转换规则：")
check("transformers/ 目录存在", dir_exists("transformers"))

# ── 汇总 ──
print("\n" + "─" * 50)
print(f"\n📊 结果：{passed} 通过 / {failed} 失败 / 共 {passed + failed} 项")

if failed > 0:
    print("\n⚠️  部分检查未通过，请根据上述提示修复。\n")
    sys.exit(1)
else:
    print("\n🎉 所有检查通过！项目结构完整。\n")
    sys.exit(0)
